tests/torchtune/models/test_lora_llama2.py

```diff
@@ -171,3 +180,100 @@ def test_lora_llama2_state_dict_parity(
+                embed_dim=512,
+            )
+            qlora = self.get_lora_llama2(
+                lora_modules=["q_proj", "v_proj", "k_proj", "output_proj"],
+                apply_lora_to_mlp=True,
+                apply_lora_to_output=False,
+                vocab_size=50,
+                quantize_base=True,
+                embed_dim=512,
+            )
+        qlora_sd = qlora.state_dict()
+        model_ref.load_state_dict(qlora_sd)
+        # Forward pass of model_ref and qlora should be the same, as QLoRA linear layers should use
+        # a special linear operator that runs the compute in bf16, but only saves the 4 bit tensors
+        # for backward.
+        ref_output = model_ref(inputs)
+        output = qlora(inputs)
+        torch.testing.assert_close(ref_output, output)
+
+    def test_qlora_llama2_state_dict(self):
+        with utils.set_default_dtype(torch.bfloat16):
+            model_ref = self.get_lora_llama2(
+                lora_modules=["q_proj", "v_proj", "k_proj", "output_proj"],
+                apply_lora_to_mlp=True,
+                apply_lora_to_output=False,
+                vocab_size=50,
+                quantize_base=False,
+                embed_dim=512,
+            )
+            bf16_sd = model_ref.state_dict()
+            for v in bf16_sd.values():
+                assert v.dtype == torch.bfloat16
+
+            # ensure quantized LoRA can load a bf16 state_dict
+            qlora = self.get_lora_llama2(
+                lora_modules=["q_proj", "v_proj", "k_proj", "output_proj"],
+                apply_lora_to_mlp=True,
+                apply_lora_to_output=False,
+                vocab_size=50,
+                quantize_base=True,
+                embed_dim=512,
+            )
+            qlora.load_state_dict(bf16_sd)
+            # LoRALinear weights should be nf4 still
+            for module in qlora.modules():
+                if isinstance(module, LoRALinear):
+                    assert isinstance(module.weight, NF4Tensor)
+            # saved state_dict should have bf16 weights.
+            qlora_sd = qlora.state_dict()
+            for v in qlora_sd.values():
+                assert v.dtype == torch.bfloat16
+
+    def test_qlora_llama2_merged_state_dict(self):
+        with utils.set_default_dtype(torch.bfloat16):
+            qlora = self.get_lora_llama2(
+                lora_modules=["q_proj", "v_proj", "k_proj", "output_proj"],
+                apply_lora_to_mlp=True,
+                apply_lora_to_output=False,
+                vocab_size=50,
+                quantize_base=True,
+                embed_dim=512,
+            )
+
+        qlora_sd = qlora.state_dict()
+        # Ensure checkpoint merging produces bf16 tensors
+        merged_ckpt = get_merged_lora_ckpt(deepcopy(qlora_sd), rank=RANK, alpha=ALPHA)
+        for v in merged_ckpt.values():
+            # paranoid check for both, as NF4Tensor had issue where NF4Tensor.dtype would return bf16
```

Comment:
Is this resolved?

Reply:
Not currently - 
<img width="764" alt="image" src="https://github.com/pytorch/torchtune/assets/8039770/04857291-8d0c-4bf3-a954-42299ce9dd1b">